lib/renderers/0-3.js

```diff
@@ -0,0 +1,268 @@
+
+    return {
+      card,
+      payload
+    };
+  }
+
+  _createUnknownCard(name) {
+    return {
+      name,
+      type: RENDER_TYPE,
+      render: this.unknownCardHandler
+    };
+  }
+
+  renderCardSection([type, index]) {
+    let { card, payload } = this._findCardByIndex(index);
+
+    let cardArg = this._createCardArgument(card, payload);
+    let rendered = card.render(cardArg);
+
+    this._validateCardRender(rendered, card.name);
+
+    return rendered || '';
+  }
+
+  _validateCardRender(rendered, cardName) {
+    if (!rendered) {
+      return;
+    }
+
+    if (typeof rendered !== 'string') {
+      throw new Error(`Card "${cardName}" must render ${RENDER_TYPE}, but result was ${typeof rendered}"`);
+    }
+  }
+
+  _registerTeardownCallback(callback) {
+    this._teardownCallbacks.push(callback);
+  }
+
+  _createCardArgument(card, payload={}) {
+    let env = {
+      name: card.name,
+      isInEditor: false,
+      onTeardown: (callback) => this._registerTeardownCallback(callback)
+    };
+
+    let options = this.cardOptions;
+
+    return { env, options, payload };
+  }
+
+  renderMarkupSection([type, tagName, markers]) {
+    return this.renderMarkers(markers);
+  }
+
+  findAtom(name) {
+    for (let i=0; i < this.atoms.length; i++) {
+      if (this.atoms[i].name === name) {
+        return this.atoms[i];
+      }
+    }
+    return this._createUnknownAtom(name);
+  }
+
+  _createUnknownAtom(name) {
+    return {
+      name,
+      type: RENDER_TYPE,
+      render: this.unknownAtomHandler
+    };
+  }
+
+  _createAtomArgument(atom, value, payload) {
+    let env = {
+      name: atom.name,
+      onTeardown: (callback) => this._registerTeardownCallback(callback)
+    };
+
+    let options = this.cardOptions;
+
+    return { env, options, value, payload };
+  }
+
+  _validateAtomRender(rendered, atomName) {
+    if (!rendered) {
+      return;
+    }
+
+    if (typeof rendered !== 'string') {
+      throw new Error(`Atom "${atomName}" must render ${RENDER_TYPE}, but result was ${typeof rendered}"`);
+    }
+  }
+
+  _findAtomByIndex(index) {
+    let atomType = this.atomTypes[index];
+    if (!atomType) {
+      throw new Error(`No atom definition found at index ${index}`);
+    }
+
+    let [ name, value, payload ] = atomType;
+    let atom = this.findAtom(name);
+
+    return {
+      atom,
+      value,
+      payload
+    };
+  }
+
+  _renderAtom(index) {
+    let { atom, value, payload } = this._findAtomByIndex(index);
+
+    let atomArg = this._createAtomArgument(atom, value, payload);
+    let rendered = atom.render(atomArg);
+
+    this._validateAtomRender(rendered, atom.name);
+
+    return rendered || '';
+  }
+
+  renderMarkers(markers) {
+    let str = '';
+    markers.forEach(m => {
+      let [type, , , value] = m;
+      switch (type) {
+        case MARKUP_MARKER_TYPE:
+          str += value;
+          break;
+        case ATOM_MARKER_TYPE:
+          str += this._renderAtom(value);
```

Comment:
Is this arg supposed to be `index` instead of `value`? `_renderAtom` wants an `index`, looks like.